third_party/lua/luadebug.lua

```diff
@@ -465,13 +670,13 @@ local function cmd_up()
 end
 
 local function cmd_where(context_lines)
-    local info = debug.getinfo(stack_inspect_offset + CMD_STACK_LEVEL, "Snl")
+    local info = debug.getinfo(stack_inspect_offset + CMD_STACK_LEVEL, "Sl")
     return (info and where(info, tonumber(context_lines) or 5))
 end
```

Comment:
Once again, this is not patch specific. `where` output is a bit confusing:
```
luadebug.lua> w 1
   1   function hello_world()
   2=>     print('hello world!')
   3   end
   1   function hello_world()
   2=>     print('hello world!')
   3   end
```
Maybe introduce some delimiter between current code listing and `where` output? Or do as gdb does and list code only when asked?

Reply:
I think, that such delimiter could be a good idea for all commands. For example `bt` output mixes with code listing too:
```
luadebug.lua> bt
Inspecting frame 0
   0 => /home/ubuntu/Code/review/tarantool-tdbg/build/test.lua:2 in global 'hello_world'
   1    ./dbg.lua:8 in chunk at ./dbg.lua:0
   1   function hello_world()
   2=>     print('hello world!')
   3   end
```

In this case, it is distinct by color, but generally speaking I don't see why would I want code listing and backtrace line by line.

Reply:
> Maybe introduce some delimiter between current code listing and where output? 
> Or do as gdb does and list code only when asked?

You've raised a good question, thanks! There is no much point to show (current) context after command which showed their own context, i.e. after `where`, `up`, and `down`. (And no, I did not introduce any kind of delimiter, I just avoid redundant information to be displayed), i.e.

![image](https://user-images.githubusercontent.com/4956130/214696270-4b84fc2c-cfdc-4705-807d-2c18ecf6633a.png)

`gdb` behaviour here (when outside of TUI mode it does not show listing unless specifically asked) is much less user-friendly than `lldb`, which always shows current context, and I personally believe that's the correct approach.

Reply:
Please see below patch I've developed to make code listing to be shown in a more reasonable way:
- I've corrected `dbg.cfg.auto_where` to actually define number of lines for context shown, not simple boolean;
- 've reimplemented "auto listing" to be triggered but that `dbg.cfg.auto_where` parameter, as they are about the same;
- `listing_shown` variable used by commands to indicate they already displayed necessary context, thus auto listing is redundant on the next iteration;

```diff
diff --git a/third_party/lua/luadebug.lua b/third_party/lua/luadebug.lua
index 903ddda76..ed059b08d 100644
--- a/third_party/lua/luadebug.lua
+++ b/third_party/lua/luadebug.lua
@@ -38,7 +38,6 @@ local COLOR_RESET = ""
 local CARET_SYM = "=>"
 local CARET = " " .. CARET_SYM .. " "
 local BREAK_SYM = "●"
-local auto_listing = true
 
 local LJ_MAX_LINE = 0x7fffff00 -- Max. source code line number.
 
@@ -416,6 +415,18 @@ local function where(info, context_lines)
     return false
 end
 
+-- Status flag to avoid redundant code listing to be shown.
+-- `up`, `down`, `where` commands show their own code context
+-- so do not show anything after them.
+local listing_shown = false
+
+-- Display automatically `where` context if configured so.
+local function auto_listing(info)
+    if tonumber(dbg.cfg.auto_where) then
+        where(info, dbg.cfg.auto_where)
+    end
+end
+
 -- Wee version differences
 local unpack = unpack or table.unpack
 local pack = function(...) return { n = select("#", ...), ... } end
@@ -590,7 +601,8 @@ local function cmd_remove_breakpoint(bps)
     if removed then
         dbg_writeln("Removed breakpoint %s:%d.", color_blue(normfile), line)
     else
-        dbg_writeln("No breakpoint defined in %s:%d.", color_blue(normfile), line)
+        dbg_write_warn("No breakpoint defined in %s:%d.",
+                       color_blue(normfile), line)
     end
     return false
 end
@@ -613,7 +625,7 @@ local function cmd_list_breakpoints()
             end
         end
     else
-        dbg_writeln("No active breakpoints defined.")
+        dbg_write_warn("No active breakpoints defined.")
     end
     return false
 end
@@ -633,12 +645,11 @@ local function cmd_up()
     if info then
         stack_inspect_offset = offset
         dbg_writeln("Inspecting frame: " .. format_stack_frame_info(info))
-        if tonumber(dbg.cfg.auto_where) then
-            where(info, dbg.cfg.auto_where)
-        end
+        auto_listing(info)
     else
-        dbg_writeln("Already at the bottom of the stack.")
+        dbg_write_warn("Already at the bottom of the stack.")
     end
+    listing_shown = true
 
     return false
 end
@@ -659,25 +670,20 @@ local function cmd_down()
     if info then
         stack_inspect_offset = offset
         dbg_writeln("Inspecting frame: " .. format_stack_frame_info(info))
-        if tonumber(dbg.cfg.auto_where) then
-            where(info, dbg.cfg.auto_where)
-        end
+        auto_listing(info)
     else
-        dbg_writeln("Already at the top of the stack.")
+        dbg_write_warn("Already at the top of the stack.")
     end
+    listing_shown = true
 
     return false
 end
 
 local function cmd_where(context_lines)
     local info = debug.getinfo(stack_inspect_offset + CMD_STACK_LEVEL, "Sl")
-    return (info and where(info, tonumber(context_lines) or 5))
-end
-
-local function cmd_listing(context_lines)
-    local offset = stack_inspect_offset + CMD_STACK_LEVEL - 2
-    local info = debug.getinfo(offset, "Sl")
-    return (info and where(info, tonumber(context_lines) or 5))
+    local ret = (info and where(info, tonumber(context_lines) or 5))
+    listing_shown = true
+    return ret
 end
 
 local function cmd_trace()
@@ -901,14 +907,14 @@ repl = function(reason)
              CARET) or ""
     dbg_writeln(reason .. format_stack_frame_info(info))
 
-    if tonumber(dbg.cfg.auto_where) then
-        where(info, dbg.cfg.auto_where)
-    end
-
     repeat
-        if auto_listing then
-            pcall(cmd_listing(3))
+        -- Do not current context if prior command showed their own.
+        if not listing_shown then
+            auto_listing(info)
         end
+        -- Command could show their own context with listing
+        -- so reset status before command executed.
+        listing_shown = false
         local success, done, hook = pcall(run_command,
                                          dbg.read(color_red(DEBUGGER .. "> ")))
         if success then
@@ -932,7 +938,7 @@ dbg = setmetatable({
         exit    = function(err) os.exit(err) end,
 
         cfg = {
-            auto_where  = false,
+            auto_where  = 3,
             auto_eval   = false,
             pretty_depth = 3,
         },
```

Reply:
Thanks for the change, it's much better now! I feel that having a delim between `bt` output and listing is a good idea, but it has different enough styling to be ok for now. Consider it a non-blocker.